File: tests/fixtures/python/service.py
from repos import AuditLog, Notifier, UserRepository


class UserService:
    def __init__(self, repo: UserRepository, notifier: Notifier) -> None:
        self.repo = repo
        self.audit = AuditLog()
        self.notifier = notifier

    async def remove(self, user_id: int) -> None:
        user = self.repo.find_user(user_id)
        await self.repo.delete_user(user_id)
        self.audit.delete_user(user_id)
        self.notifier.send(f"removed {user['id']}")


class Services:
    def __init__(self, users: UserService) -> None:
        self.users = users


class App:
    def __init__(self, services: Services) -> None:
        self.services = services


async def handle_delete(app: App, user_id: int) -> None:
    await app.services.users.remove(user_id)


def cleanup(user_id: int) -> None:
    repo = AuditLog()

    async def purge() -> None:
        repo = UserRepository()
        await repo.delete_user(user_id)

    repo.delete_user(user_id)

File: tests/fixtures/typescript/service.ts
import { AuditLog, Notifier, UserRepository } from "./repos";

export class UserService {
  private audit = new AuditLog();

  constructor(
    private repo: UserRepository,
    private notifier: Notifier,
  ) {}

  async remove(id: number): Promise<void> {
    const user = this.repo.findUser(id);
    await this.repo.deleteUser(id);
    this.audit.deleteUser(id);
    this.notifier.send(`removed ${user.id}`);
  }
}

export class Services {
  constructor(public users: UserService) {}
}

export class App {
  constructor(public services: Services) {}
}

export async function handleDelete(app: App, id: number): Promise<void> {
  await app.services.users.remove(id);
}

export function cleanup(id: number): void {
  const repo = new AuditLog();
  const purge = async () => {
    const repo = new UserRepository();
    await repo.deleteUser(id);
  };
  void purge;
  repo.deleteUser(id);
}
